From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Mon, 12 Oct 2026 22:13:43 -0700
Subject: [PATCH] mshtml: markup pointers and the editing of IMarkupServices

IMarkupPointer/IMarkupPointer2 were all E_NOTIMPL; they are positions between
the tokens of a document (begin tag, end tag, each character of text) kept as
collapsed DOM ranges, so they follow the document as it changes. Implemented:
MoveAdjacentToElement, MoveToPointer, MoveToContainer, Left and Right over
text, begin and end tags and no-scope elements, CurrentScope, the five
comparisons, GetContainer, OwningDoc, gravity and cling, Unposition,
IsPositioned, MoveUnit for characters and words, FindText, IsAtWordBreak,
GetMarkupPosition/MoveToMarkupPosition, MoveUnitBounded, MoveToContent.
IMarkupServices: CreateElement (tag ids and attribute text), CloneElement,
InsertElement, RemoveElement, Remove, Copy, Move, InsertText, ParseGlobal,
ParseString with pointers, IsScopedElement, the tag id and name lookups,
MovePointersToRange, MoveRangeToPointers, undo units, and
IMarkupContainer::OwningDoc.

diff --git a/dlls/mshtml/htmldoc.c b/dlls/mshtml/htmldoc.c
index 988eb0f..682e9a2 100644
--- a/dlls/mshtml/htmldoc.c
+++ b/dlls/mshtml/htmldoc.c
@@ -5042,7 +5042,7 @@ static HRESULT WINAPI MarkupServices_CreateMarkupPointer(IMarkupServices *iface,
 
     TRACE("(%p)->(%p)\n", This, ppPointer);
 
-    return create_markup_pointer(ppPointer);
+    return create_markup_pointer(This, ppPointer);
 }
 
 static HRESULT WINAPI MarkupServices_CreateMarkupContainer(IMarkupServices *iface, IMarkupContainer **ppMarkupContainer)
@@ -5062,20 +5062,229 @@ static HRESULT WINAPI MarkupServices_CreateMarkupContainer(IMarkupServices *ifac
     return S_OK;
 }
 
+/* the tags that IMarkupServices knows by number */
+static const struct { ELEMENT_TAG_ID id; const WCHAR *name; } markup_tags[] = {
+    { TAGID_A, L"a" },
+    { TAGID_ACRONYM, L"acronym" },
+    { TAGID_ADDRESS, L"address" },
+    { TAGID_APPLET, L"applet" },
+    { TAGID_AREA, L"area" },
+    { TAGID_B, L"b" },
+    { TAGID_BASE, L"base" },
+    { TAGID_BASEFONT, L"basefont" },
+    { TAGID_BDO, L"bdo" },
+    { TAGID_BGSOUND, L"bgsound" },
+    { TAGID_BIG, L"big" },
+    { TAGID_BLINK, L"blink" },
+    { TAGID_BLOCKQUOTE, L"blockquote" },
+    { TAGID_BODY, L"body" },
+    { TAGID_BR, L"br" },
+    { TAGID_BUTTON, L"button" },
+    { TAGID_CAPTION, L"caption" },
+    { TAGID_CENTER, L"center" },
+    { TAGID_CITE, L"cite" },
+    { TAGID_CODE, L"code" },
+    { TAGID_COL, L"col" },
+    { TAGID_COLGROUP, L"colgroup" },
+    { TAGID_DD, L"dd" },
+    { TAGID_DEL, L"del" },
+    { TAGID_DFN, L"dfn" },
+    { TAGID_DIR, L"dir" },
+    { TAGID_DIV, L"div" },
+    { TAGID_DL, L"dl" },
+    { TAGID_DT, L"dt" },
+    { TAGID_EM, L"em" },
+    { TAGID_EMBED, L"embed" },
+    { TAGID_FIELDSET, L"fieldset" },
+    { TAGID_FONT, L"font" },
+    { TAGID_FORM, L"form" },
+    { TAGID_FRAME, L"frame" },
+    { TAGID_FRAMESET, L"frameset" },
+    { TAGID_H1, L"h1" },
+    { TAGID_H2, L"h2" },
+    { TAGID_H3, L"h3" },
+    { TAGID_H4, L"h4" },
+    { TAGID_H5, L"h5" },
+    { TAGID_H6, L"h6" },
+    { TAGID_HEAD, L"head" },
+    { TAGID_HR, L"hr" },
+    { TAGID_HTML, L"html" },
+    { TAGID_I, L"i" },
+    { TAGID_IFRAME, L"iframe" },
+    { TAGID_IMG, L"img" },
+    { TAGID_INPUT, L"input" },
+    { TAGID_INS, L"ins" },
+    { TAGID_KBD, L"kbd" },
+    { TAGID_LABEL, L"label" },
+    { TAGID_LEGEND, L"legend" },
+    { TAGID_LI, L"li" },
+    { TAGID_LINK, L"link" },
+    { TAGID_LISTING, L"listing" },
+    { TAGID_MAP, L"map" },
+    { TAGID_MARQUEE, L"marquee" },
+    { TAGID_MENU, L"menu" },
+    { TAGID_META, L"meta" },
+    { TAGID_NEXTID, L"nextid" },
+    { TAGID_NOBR, L"nobr" },
+    { TAGID_NOEMBED, L"noembed" },
+    { TAGID_NOFRAMES, L"noframes" },
+    { TAGID_NOSCRIPT, L"noscript" },
+    { TAGID_OBJECT, L"object" },
+    { TAGID_OL, L"ol" },
+    { TAGID_OPTION, L"option" },
+    { TAGID_P, L"p" },
+    { TAGID_PARAM, L"param" },
+    { TAGID_PLAINTEXT, L"plaintext" },
+    { TAGID_PRE, L"pre" },
+    { TAGID_Q, L"q" },
+    { TAGID_RP, L"rp" },
+    { TAGID_RT, L"rt" },
+    { TAGID_RUBY, L"ruby" },
+    { TAGID_S, L"s" },
+    { TAGID_SAMP, L"samp" },
+    { TAGID_SCRIPT, L"script" },
+    { TAGID_SELECT, L"select" },
+    { TAGID_SMALL, L"small" },
+    { TAGID_SPAN, L"span" },
+    { TAGID_STRIKE, L"strike" },
+    { TAGID_STRONG, L"strong" },
+    { TAGID_STYLE, L"style" },
+    { TAGID_SUB, L"sub" },
+    { TAGID_SUP, L"sup" },
+    { TAGID_TABLE, L"table" },
+    { TAGID_TBODY, L"tbody" },
+    { TAGID_TC, L"tc" },
+    { TAGID_TD, L"td" },
+    { TAGID_TEXTAREA, L"textarea" },
+    { TAGID_TFOOT, L"tfoot" },
+    { TAGID_TH, L"th" },
+    { TAGID_THEAD, L"thead" },
+    { TAGID_TITLE, L"title" },
+    { TAGID_TR, L"tr" },
+    { TAGID_TT, L"tt" },
+    { TAGID_U, L"u" },
+    { TAGID_UL, L"ul" },
+    { TAGID_VAR, L"var" },
+    { TAGID_WBR, L"wbr" },
+    { TAGID_XMP, L"xmp" },
+    { TAGID_OPTGROUP, L"optgroup" },
+    { TAGID_ABBR, L"abbr" },
+};
+
+static const WCHAR *markup_tag_name(ELEMENT_TAG_ID id)
+{
+    unsigned i;
+    for(i = 0; i < ARRAY_SIZE(markup_tags); i++)
+        if(markup_tags[i].id == id)
+            return markup_tags[i].name;
+    return NULL;
+}
+
+static ELEMENT_TAG_ID markup_tag_id(const WCHAR *name)
+{
+    unsigned i;
+    for(i = 0; i < ARRAY_SIZE(markup_tags); i++)
+        if(!wcsicmp(markup_tags[i].name, name))
+            return markup_tags[i].id;
+    return TAGID_UNKNOWN;
+}
+
+/* name=value pairs, the value bare or in quotes */
+static HRESULT apply_markup_attributes(IHTMLElement *elem, const WCHAR *attrs)
+{
+    HRESULT hres = S_OK;
+
+    while(attrs && *attrs && SUCCEEDED(hres)) {
+        const WCHAR *name, *value;
+        size_t name_len, value_len;
+        WCHAR quote = 0;
+        BSTR bname, bvalue;
+        VARIANT v;
+
+        while(iswspace(*attrs)) attrs++;
+        if(!*attrs)
+            break;
+        name = attrs;
+        while(*attrs && *attrs != '=' && !iswspace(*attrs)) attrs++;
+        name_len = attrs - name;
+        while(iswspace(*attrs)) attrs++;
+        value = attrs;
+        value_len = 0;
+        if(*attrs == '=') {
+            attrs++;
+            while(iswspace(*attrs)) attrs++;
+            if(*attrs == '"' || *attrs == '\'')
+                quote = *attrs++;
+            value = attrs;
+            while(*attrs && (quote ? *attrs != quote : !iswspace(*attrs))) attrs++;
+            value_len = attrs - value;
+            if(quote && *attrs == quote)
+                attrs++;
+        }else {
+            value = L"";
+        }
+        if(!name_len)
+            return E_INVALIDARG;
+
+        bname = SysAllocStringLen(name, name_len);
+        bvalue = SysAllocStringLen(value, value_len);
+        V_VT(&v) = VT_BSTR;
+        V_BSTR(&v) = bvalue;
+        hres = IHTMLElement_setAttribute(elem, bname, v, 0);
+        SysFreeString(bname);
+        SysFreeString(bvalue);
+    }
+    return hres;
+}
+
 static HRESULT WINAPI MarkupServices_CreateElement(IMarkupServices *iface,
     ELEMENT_TAG_ID tagID, OLECHAR *pchAttributes, IHTMLElement **ppElement)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%d,%s,%p)\n", This, tagID, debugstr_w(pchAttributes), ppElement);
-    return E_NOTIMPL;
+    const WCHAR *name;
+    BSTR bname;
+    HRESULT hres;
+
+    TRACE("(%p)->(%d,%s,%p)\n", This, tagID, debugstr_w(pchAttributes), ppElement);
+
+    if(!(name = markup_tag_name(tagID)))
+        return E_INVALIDARG;
+    if(!(bname = SysAllocString(name)))
+        return E_OUTOFMEMORY;
+    hres = IHTMLDocument2_createElement(&This->IHTMLDocument2_iface, bname, ppElement);
+    SysFreeString(bname);
+    if(SUCCEEDED(hres)) {
+        hres = apply_markup_attributes(*ppElement, pchAttributes);
+        if(FAILED(hres)) {
+            IHTMLElement_Release(*ppElement);
+            *ppElement = NULL;
+        }
+    }
+    return hres;
 }
 
 static HRESULT WINAPI MarkupServices_CloneElement(IMarkupServices *iface,
     IHTMLElement *pElemCloneThis, IHTMLElement **ppElementTheClone)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%p,%p)\n", This, pElemCloneThis, ppElementTheClone);
-    return E_NOTIMPL;
+    IHTMLDOMNode *node, *clone;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p,%p)\n", This, pElemCloneThis, ppElementTheClone);
+
+    if(!pElemCloneThis)
+        return E_INVALIDARG;
+    hres = IHTMLElement_QueryInterface(pElemCloneThis, &IID_IHTMLDOMNode, (void**)&node);
+    if(FAILED(hres))
+        return hres;
+    /* the clone has the attributes of the element, not its content */
+    hres = IHTMLDOMNode_cloneNode(node, VARIANT_FALSE, &clone);
+    IHTMLDOMNode_Release(node);
+    if(FAILED(hres))
+        return hres;
+    hres = IHTMLDOMNode_QueryInterface(clone, &IID_IHTMLElement, (void**)ppElementTheClone);
+    IHTMLDOMNode_Release(clone);
+    return hres;
 }
 
 static HRESULT WINAPI MarkupServices_InsertElement(IMarkupServices *iface,
@@ -5083,23 +5292,29 @@ static HRESULT WINAPI MarkupServices_InsertElement(IMarkupServices *iface,
     IMarkupPointer *pPointerFinish)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%p,%p,%p)\n", This, pElementInsert, pPointerStart, pPointerFinish);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p,%p,%p)\n", This, pElementInsert, pPointerStart, pPointerFinish);
+
+    return markup_insert_element(pElementInsert, pPointerStart, pPointerFinish);
 }
 
 static HRESULT WINAPI MarkupServices_RemoveElement(IMarkupServices *iface, IHTMLElement *pElementRemove)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%p)\n", This, pElementRemove);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, pElementRemove);
+
+    return markup_remove_element(pElementRemove);
 }
 
 static HRESULT WINAPI MarkupServices_Remove(IMarkupServices *iface,
     IMarkupPointer *pPointerStart, IMarkupPointer *pPointerFinish)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%p,%p)\n", This, pPointerStart, pPointerFinish);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p,%p)\n", This, pPointerStart, pPointerFinish);
+
+    return markup_remove(pPointerStart, pPointerFinish);
 }
 
 static HRESULT WINAPI MarkupServices_Copy(IMarkupServices *iface,
@@ -5107,8 +5322,10 @@ static HRESULT WINAPI MarkupServices_Copy(IMarkupServices *iface,
     IMarkupPointer *pPointerTarget)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%p,%p,%p)\n", This, pPointerSourceStart, pPointerSourceFinish, pPointerTarget);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p,%p,%p)\n", This, pPointerSourceStart, pPointerSourceFinish, pPointerTarget);
+
+    return markup_copy(pPointerSourceStart, pPointerSourceFinish, pPointerTarget, FALSE);
 }
 
 static HRESULT WINAPI MarkupServices_Move(IMarkupServices *iface,
@@ -5116,16 +5333,20 @@ static HRESULT WINAPI MarkupServices_Move(IMarkupServices *iface,
     IMarkupPointer *pPointerTarget)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%p,%p,%p)\n", This, pPointerSourceStart, pPointerSourceFinish, pPointerTarget);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p,%p,%p)\n", This, pPointerSourceStart, pPointerSourceFinish, pPointerTarget);
+
+    return markup_copy(pPointerSourceStart, pPointerSourceFinish, pPointerTarget, TRUE);
 }
 
 static HRESULT WINAPI MarkupServices_InsertText(IMarkupServices *iface,
     OLECHAR *pchText, LONG cch, IMarkupPointer *pPointerTarget)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%s,%lx,%p)\n", This, debugstr_w(pchText), cch, pPointerTarget);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s,%lx,%p)\n", This, debugstr_w(pchText), cch, pPointerTarget);
+
+    return markup_insert_text(pchText, cch, pPointerTarget);
 }
 
 static HRESULT WINAPI MarkupServices_ParseString(IMarkupServices *iface,
@@ -5142,11 +5363,6 @@ static HRESULT WINAPI MarkupServices_ParseString(IMarkupServices *iface,
 
     if(dwFlags != 0)
         FIXME("flags %lx not implemented.\n", dwFlags);
-    if(pPointerStart || pPointerFinish) {
-        FIXME("Pointers not implemented.\n");
-        return E_NOTIMPL;
-    }
-
     hres = IMarkupServices_CreateMarkupContainer(iface, &container);
     if(FAILED(hres))
         return hres;
@@ -5173,6 +5389,14 @@ static HRESULT WINAPI MarkupServices_ParseString(IMarkupServices *iface,
         container = NULL;
     }
 
+    if(SUCCEEDED(hres)) {
+        /* the pointers get the ends of what was parsed */
+        if(pPointerStart)
+            hres = IMarkupPointer_MoveAdjacentToElement(pPointerStart, body, ELEM_ADJ_AfterBegin);
+        if(SUCCEEDED(hres) && pPointerFinish)
+            hres = IMarkupPointer_MoveAdjacentToElement(pPointerFinish, body, ELEM_ADJ_BeforeEnd);
+    }
+
     IHTMLDocument2_Release(doc);
     IHTMLElement_Release(body);
 
@@ -5186,70 +5410,115 @@ static HRESULT WINAPI MarkupServices_ParseGlobal(IMarkupServices *iface,
     IMarkupPointer *pPointerStart, IMarkupPointer *pPointerFinish)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%s,%lx,%p,%p,%p)\n", This, debugstr_w(hglobalHTML), dwFlags, ppContainerResult, pPointerStart, pPointerFinish);
-    return E_NOTIMPL;
+    WCHAR *html;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p,%lx,%p,%p,%p)\n", This, hglobalHTML, dwFlags, ppContainerResult, pPointerStart, pPointerFinish);
+
+    if(!hglobalHTML)
+        return E_INVALIDARG;
+    /* the memory holds the text as a null-terminated wide string */
+    if(!(html = GlobalLock(hglobalHTML)))
+        return E_INVALIDARG;
+    hres = IMarkupServices_ParseString(iface, html, dwFlags, ppContainerResult, pPointerStart, pPointerFinish);
+    GlobalUnlock(hglobalHTML);
+    return hres;
 }
 
 static HRESULT WINAPI MarkupServices_IsScopedElement(IMarkupServices *iface,
     IHTMLElement *pElement, BOOL *pfScoped)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%p,%p)\n", This, pElement, pfScoped);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p,%p)\n", This, pElement, pfScoped);
+
+    if(!pElement)
+        return E_INVALIDARG;
+    return markup_is_scoped(pElement, pfScoped);
 }
 
 static HRESULT WINAPI MarkupServices_GetElementTagId(IMarkupServices *iface,
     IHTMLElement *pElement, ELEMENT_TAG_ID *ptagId)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%p,%p)\n", This, pElement, ptagId);
-    return E_NOTIMPL;
+    BSTR name;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p,%p)\n", This, pElement, ptagId);
+
+    if(!pElement)
+        return E_INVALIDARG;
+    hres = IHTMLElement_get_tagName(pElement, &name);
+    if(FAILED(hres))
+        return hres;
+    *ptagId = name ? markup_tag_id(name) : TAGID_UNKNOWN;
+    SysFreeString(name);
+    return S_OK;
 }
 
 static HRESULT WINAPI MarkupServices_GetTagIDForName(IMarkupServices *iface,
     BSTR bstrName, ELEMENT_TAG_ID *ptagId)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%s,%p)\n", This, debugstr_w(bstrName), ptagId);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s,%p)\n", This, debugstr_w(bstrName), ptagId);
+
+    if(!bstrName)
+        return E_INVALIDARG;
+    *ptagId = markup_tag_id(bstrName);
+    return S_OK;
 }
 
 static HRESULT WINAPI MarkupServices_GetNameForTagID(IMarkupServices *iface,
     ELEMENT_TAG_ID tagId, BSTR *pbstrName)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%d,%p)\n", This, tagId, pbstrName);
-    return E_NOTIMPL;
+    const WCHAR *name;
+
+    TRACE("(%p)->(%d,%p)\n", This, tagId, pbstrName);
+
+    if(!(name = markup_tag_name(tagId)))
+        return E_INVALIDARG;
+    return (*pbstrName = SysAllocString(name)) ? S_OK : E_OUTOFMEMORY;
 }
 
 static HRESULT WINAPI MarkupServices_MovePointersToRange(IMarkupServices *iface,
     IHTMLTxtRange *pIRange, IMarkupPointer *pPointerStart, IMarkupPointer *pPointerFinish)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%p,%p,%p)\n", This, pIRange, pPointerStart, pPointerFinish);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p,%p,%p)\n", This, pIRange, pPointerStart, pPointerFinish);
+
+    return markup_pointers_to_range(pIRange, pPointerStart, pPointerFinish);
 }
 
 static HRESULT WINAPI MarkupServices_MoveRangeToPointers(IMarkupServices *iface,
     IMarkupPointer *pPointerStart, IMarkupPointer *pPointerFinish, IHTMLTxtRange *pIRange)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%p,%p,%p)\n", This, pPointerStart, pPointerFinish, pIRange);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p,%p,%p)\n", This, pPointerStart, pPointerFinish, pIRange);
+
+    return markup_range_to_pointers(pPointerStart, pPointerFinish, pIRange);
 }
 
 static HRESULT WINAPI MarkupServices_BeginUndoUnit(IMarkupServices *iface, OLECHAR *pchTitle)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(pchTitle));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(pchTitle));
+
+    /* there is no undo stack to group the changes in */
+    return S_OK;
 }
 
 static HRESULT WINAPI MarkupServices_EndUndoUnit(IMarkupServices *iface)
 {
     HTMLDocumentNode *This = impl_from_IMarkupServices(iface);
-    FIXME("(%p)\n", This);
-    return E_NOTIMPL;
+
+    TRACE("(%p)\n", This);
+
+    return S_OK;
 }
 
 static const IMarkupServicesVtbl MarkupServicesVtbl = {
@@ -5307,10 +5576,14 @@ static ULONG WINAPI MarkupContainer_Release(IMarkupContainer *iface)
 static HRESULT WINAPI MarkupContainer_OwningDoc(IMarkupContainer *iface, IHTMLDocument2 **ppDoc)
 {
     HTMLDocumentNode *This = impl_from_IMarkupContainer(iface);
-    FIXME("(%p)->(%p)\n", This, ppDoc);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, ppDoc);
+
+    IHTMLDocument2_AddRef(*ppDoc = &This->IHTMLDocument2_iface);
+    return S_OK;
 }
 
+
 static const IMarkupContainerVtbl MarkupContainerVtbl = {
     MarkupContainer_QueryInterface,
     MarkupContainer_AddRef,
@@ -5318,6 +5591,11 @@ static const IMarkupContainerVtbl MarkupContainerVtbl = {
     MarkupContainer_OwningDoc
 };
 
+HTMLDocumentNode *unsafe_impl_from_IMarkupContainer(IMarkupContainer *iface)
+{
+    return iface->lpVtbl == &MarkupContainerVtbl ? impl_from_IMarkupContainer(iface) : NULL;
+}
+
 /**********************************************************
  * IDisplayServices implementation
  */
diff --git a/dlls/mshtml/mshtml_private.h b/dlls/mshtml/mshtml_private.h
index 0abc075..2d2bf54 100644
--- a/dlls/mshtml/mshtml_private.h
+++ b/dlls/mshtml/mshtml_private.h
@@ -1301,7 +1301,16 @@ HRESULT create_style_sheet_collection(nsIDOMStyleSheetList*,HTMLDocumentNode*,
 HRESULT create_dom_range(nsIDOMRange*,HTMLDocumentNode*,IHTMLDOMRange**);
 HRESULT create_html_rect(nsIDOMClientRect*,DispatchEx*,IHTMLRect**);
 HRESULT create_html_rect_collection(nsIDOMClientRectList*,HTMLDocumentNode*,compat_mode_t,IHTMLRectCollection**);
-HRESULT create_markup_pointer(IMarkupPointer**);
+HRESULT create_markup_pointer(HTMLDocumentNode*,IMarkupPointer**);
+HRESULT markup_remove(IMarkupPointer*,IMarkupPointer*);
+HRESULT markup_copy(IMarkupPointer*,IMarkupPointer*,IMarkupPointer*,BOOL);
+HRESULT markup_insert_text(const WCHAR*,LONG,IMarkupPointer*);
+HRESULT markup_insert_element(IHTMLElement*,IMarkupPointer*,IMarkupPointer*);
+HRESULT markup_remove_element(IHTMLElement*);
+HRESULT markup_is_scoped(IHTMLElement*,BOOL*);
+HRESULT markup_pointers_to_range(IHTMLTxtRange*,IMarkupPointer*,IMarkupPointer*);
+HRESULT markup_range_to_pointers(IMarkupPointer*,IMarkupPointer*,IHTMLTxtRange*);
+HTMLDocumentNode *unsafe_impl_from_IMarkupContainer(IMarkupContainer*);
 
 void detach_document_node(HTMLDocumentNode*);
 void detach_selection(HTMLDocumentNode*);
diff --git a/dlls/mshtml/oleobj.c b/dlls/mshtml/oleobj.c
index c488811..bc1e690 100644
--- a/dlls/mshtml/oleobj.c
+++ b/dlls/mshtml/oleobj.c
@@ -3057,7 +3057,7 @@ static HRESULT WINAPI DocObjMarkupServices_CreateMarkupPointer(IMarkupServices *
 
     TRACE("(%p)->(%p)\n", This, ppPointer);
 
-    return create_markup_pointer(ppPointer);
+    return create_markup_pointer(This->doc_node, ppPointer);
 }
 
 HTMLDOCUMENTOBJ_FWD_TO_NODE_1(MarkupServices, CreateMarkupContainer, IMarkupContainer**)
diff --git a/dlls/mshtml/range.c b/dlls/mshtml/range.c
index 472fdad..33387b2 100644
--- a/dlls/mshtml/range.c
+++ b/dlls/mshtml/range.c
@@ -2568,16 +2568,170 @@ void detach_ranges(HTMLDocumentNode *This)
     }
 }
 
+/* Markup pointers.
+ *
+ * A markup pointer sits between the tokens of a document: the begin tag and the end tag of an
+ * element are a token each, and so is every character of text. It is kept as a collapsed DOM
+ * range, so the engine moves it along when the document changes. The positions "after the
+ * last character of a text node" and "after the text node" are one markup position, so they
+ * are compared after being moved out of the text node. */
+
 typedef struct {
     IMarkupPointer2 IMarkupPointer2_iface;
     LONG ref;
+    HTMLDocumentNode *doc;
+    nsIDOMRange *range;     /* NULL: not positioned */
+    POINTER_GRAVITY gravity;
+    BOOL cling;
 } MarkupPointer;
 
+static const IMarkupPointer2Vtbl MarkupPointer2Vtbl;
+
 static inline MarkupPointer *impl_from_IMarkupPointer2(IMarkupPointer2 *iface)
 {
     return CONTAINING_RECORD(iface, MarkupPointer, IMarkupPointer2_iface);
 }
 
+static MarkupPointer *unsafe_markup_pointer(IMarkupPointer *iface)
+{
+    return iface && iface->lpVtbl == (const IMarkupPointerVtbl*)&MarkupPointer2Vtbl
+        ? impl_from_IMarkupPointer2((IMarkupPointer2*)iface) : NULL;
+}
+
+static BOOL is_scopeless_node(nsIDOMNode *node)
+{
+    /* elements that have no end tag and no content */
+    static const WCHAR * const tags[] = { L"area", L"base", L"basefont", L"bgsound", L"br", L"col", L"embed",
+        L"frame", L"hr", L"img", L"input", L"isindex", L"link", L"meta", L"nextid", L"param", L"wbr" };
+    unsigned i;
+
+#ifdef SG_MUTANT_MSHTML_MARKUP_NOVOID
+    return FALSE;
+#endif
+    for(i = 0; i < ARRAY_SIZE(tags); i++)
+        if(is_elem_tag(node, tags[i]))
+            return TRUE;
+    return FALSE;
+}
+
+static BOOL is_container_node(nsIDOMNode *node)
+{
+    UINT16 type = get_node_type(node);
+    return type == DOCUMENT_NODE || type == DOCUMENT_FRAGMENT_NODE;
+}
+
+/* the position, with a reference on the node */
+static BOOL mp_get(MarkupPointer *This, nsIDOMNode **node, LONG *off)
+{
+    if(!This->range)
+        return FALSE;
+    nsIDOMRange_GetStartContainer(This->range, node);
+    nsIDOMRange_GetStartOffset(This->range, off);
+    return *node != NULL;
+}
+
+static HRESULT mp_set(MarkupPointer *This, nsIDOMNode *node, LONG off)
+{
+    nsIDOMDocument *nsdoc = NULL;
+    nsIDOMRange *range = NULL;
+    nsresult nsres;
+
+    if(get_node_type(node) == DOCUMENT_NODE)
+        nsIDOMNode_QueryInterface(node, &IID_nsIDOMDocument, (void**)&nsdoc);
+    else
+        nsIDOMNode_GetOwnerDocument(node, &nsdoc);
+    if(!nsdoc)
+        return E_FAIL;
+
+    nsres = nsIDOMDocument_CreateRange(nsdoc, &range);
+    nsIDOMDocument_Release(nsdoc);
+    if(NS_FAILED(nsres) || !range)
+        return E_FAIL;
+
+    nsres = nsIDOMRange_SetStart(range, node, off);
+    if(NS_SUCCEEDED(nsres))
+        nsres = nsIDOMRange_SetEnd(range, node, off);
+    if(NS_FAILED(nsres)) {
+        nsIDOMRange_Release(range);
+        return E_INVALIDARG;
+    }
+
+    if(This->range)
+        nsIDOMRange_Release(This->range);
+    This->range = range;
+    return S_OK;
+}
+
+/* a text position at either end of its node is the position next to the node */
+static void mp_normalize(nsIDOMNode **node, LONG *off)
+{
+    nsIDOMNode *parent = NULL;
+
+#ifdef SG_MUTANT_MSHTML_MARKUP_NONORM
+    return;
+#endif
+    if(get_node_type(*node) != TEXT_NODE)
+        return;
+    if(*off != 0 && *off < (LONG)get_text_length(*node))
+        return;
+
+    nsIDOMNode_GetParentNode(*node, &parent);
+    if(!parent)
+        return;
+    *off = get_child_index(parent, *node) + (*off ? 1 : 0);
+    nsIDOMNode_Release(*node);
+    *node = parent;
+}
+
+/* -1, 0 or 1; the pointers have to be positioned in the same tree */
+static HRESULT mp_compare(MarkupPointer *a, MarkupPointer *b, int *ret)
+{
+    nsIDOMNode *na, *nb;
+    LONG oa, ob;
+    nsIDOMRange *ra = NULL, *rb = NULL;
+    short cmp = 0;
+    nsresult nsres;
+    HRESULT hres = E_FAIL;
+
+    if(!a->range || !b->range)
+        return E_FAIL;
+
+    mp_get(a, &na, &oa);
+    mp_get(b, &nb, &ob);
+    mp_normalize(&na, &oa);
+    mp_normalize(&nb, &ob);
+
+    if(NS_SUCCEEDED(nsIDOMRange_CloneRange(a->range, &ra)) && NS_SUCCEEDED(nsIDOMRange_CloneRange(b->range, &rb))
+       && NS_SUCCEEDED(nsIDOMRange_SetStart(ra, na, oa)) && NS_SUCCEEDED(nsIDOMRange_SetEnd(ra, na, oa))
+       && NS_SUCCEEDED(nsIDOMRange_SetStart(rb, nb, ob)) && NS_SUCCEEDED(nsIDOMRange_SetEnd(rb, nb, ob))) {
+        nsres = nsIDOMRange_CompareBoundaryPoints(ra, NS_START_TO_START, rb, &cmp);
+        if(NS_SUCCEEDED(nsres)) {
+            *ret = cmp < 0 ? -1 : cmp > 0 ? 1 : 0;
+            hres = S_OK;
+        }else {
+            hres = E_INVALIDARG;
+        }
+    }
+    if(ra) nsIDOMRange_Release(ra);
+    if(rb) nsIDOMRange_Release(rb);
+    nsIDOMNode_Release(na);
+    nsIDOMNode_Release(nb);
+    return hres;
+}
+
+static HRESULT mp_node_to_element(nsIDOMNode *nsnode, IHTMLElement **ret)
+{
+    HTMLDOMNode *node;
+    HRESULT hres;
+
+    hres = get_node(nsnode, TRUE, &node);
+    if(FAILED(hres))
+        return hres;
+    hres = IHTMLDOMNode_QueryInterface(&node->IHTMLDOMNode_iface, &IID_IHTMLElement, (void**)ret);
+    node_release(node);
+    return hres;
+}
+
 static HRESULT WINAPI MarkupPointer2_QueryInterface(IMarkupPointer2 *iface, REFIID riid, void **ppv)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
@@ -2617,8 +2771,12 @@ static ULONG WINAPI MarkupPointer2_Release(IMarkupPointer2 *iface)
 
     TRACE("(%p) ref=%ld\n", This, ref);
 
-    if(!ref)
+    if(!ref) {
+        if(This->range)
+            nsIDOMRange_Release(This->range);
+        IHTMLDOMNode_Release(&This->doc->node.IHTMLDOMNode_iface);
         free(This);
+    }
 
     return ref;
 }
@@ -2626,179 +2784,759 @@ static ULONG WINAPI MarkupPointer2_Release(IMarkupPointer2 *iface)
 static HRESULT WINAPI MarkupPointer2_OwningDoc(IMarkupPointer2 *iface, IHTMLDocument2 **p)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    IHTMLDocument2_AddRef(*p = &This->doc->IHTMLDocument2_iface);
+    return S_OK;
 }
 
 static HRESULT WINAPI MarkupPointer2_Gravity(IMarkupPointer2 *iface, POINTER_GRAVITY *p)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    *p = This->gravity;
+    return S_OK;
 }
 
 static HRESULT WINAPI MarkupPointer2_SetGravity(IMarkupPointer2 *iface, POINTER_GRAVITY gravity)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%u)\n", This, gravity);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%u)\n", This, gravity);
+
+    if(gravity != POINTER_GRAVITY_Left && gravity != POINTER_GRAVITY_Right)
+        return E_INVALIDARG;
+    This->gravity = gravity;
+    return S_OK;
 }
 
 static HRESULT WINAPI MarkupPointer2_Cling(IMarkupPointer2 *iface, BOOL *p)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    *p = This->cling;
+    return S_OK;
 }
 
 static HRESULT WINAPI MarkupPointer2_SetCling(IMarkupPointer2 *iface, BOOL cling)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%x)\n", This, cling);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%x)\n", This, cling);
+
+    This->cling = !!cling;
+    return S_OK;
 }
 
 static HRESULT WINAPI MarkupPointer2_Unposition(IMarkupPointer2 *iface)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)\n", This);
-    return E_NOTIMPL;
+
+    TRACE("(%p)\n", This);
+
+    if(This->range) {
+        nsIDOMRange_Release(This->range);
+        This->range = NULL;
+    }
+    return S_OK;
 }
 
 static HRESULT WINAPI MarkupPointer2_IsPositioned(IMarkupPointer2 *iface, BOOL *p)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    *p = This->range != NULL;
+    return S_OK;
+}
+
+/* the document node that holds the position */
+static HRESULT mp_get_container_doc(MarkupPointer *This, HTMLDocumentNode **ret)
+{
+    nsIDOMNode *node, *top = NULL;
+    HTMLDOMNode *doc_node;
+    LONG off;
+    HRESULT hres;
+
+    if(!mp_get(This, &node, &off))
+        return E_FAIL;
+
+    /* a node outside of any document is a tree of its own: the owner document is the container */
+    if(get_node_type(node) == DOCUMENT_NODE) {
+        top = node;
+    }else {
+        nsIDOMDocument *nsdoc = NULL;
+        nsIDOMNode_GetOwnerDocument(node, &nsdoc);
+        nsIDOMNode_Release(node);
+        if(!nsdoc)
+            return E_FAIL;
+        nsIDOMDocument_QueryInterface(nsdoc, &IID_nsIDOMNode, (void**)&top);
+        nsIDOMDocument_Release(nsdoc);
+    }
+
+    hres = get_node(top, TRUE, &doc_node);
+    nsIDOMNode_Release(top);
+    if(FAILED(hres))
+        return hres;
+
+    *ret = CONTAINING_RECORD(doc_node, HTMLDocumentNode, node);
+    return S_OK;
 }
 
 static HRESULT WINAPI MarkupPointer2_GetContainer(IMarkupPointer2 *iface, IMarkupContainer **p)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    HTMLDocumentNode *doc;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    *p = NULL;
+    if(!This->range)
+        return S_OK;
+
+    hres = mp_get_container_doc(This, &doc);
+    if(FAILED(hres))
+        return hres;
+
+    hres = IHTMLDOMNode_QueryInterface(&doc->node.IHTMLDOMNode_iface, &IID_IMarkupContainer, (void**)p);
+    node_release(&doc->node);
+    return hres;
 }
 
 static HRESULT WINAPI MarkupPointer2_MoveAdjacentToElement(IMarkupPointer2 *iface, IHTMLElement *element, ELEMENT_ADJACENCY adj)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p %u)\n", This, element, adj);
-    return E_NOTIMPL;
+    nsIDOMNode *parent = NULL;
+    HTMLElement *elem;
+    HRESULT hres = S_OK;
+
+    TRACE("(%p)->(%p %u)\n", This, element, adj);
+
+    if(!element)
+        return E_INVALIDARG;
+    if(!(elem = unsafe_impl_from_IHTMLElement(element)))
+        return E_INVALIDARG;
+
+    switch(adj) {
+    case ELEM_ADJ_BeforeBegin:
+    case ELEM_ADJ_AfterEnd:
+        nsIDOMNode_GetParentNode(elem->node.nsnode, &parent);
+        if(!parent)
+            return E_INVALIDARG;
+        hres = mp_set(This, parent, get_child_index(parent, elem->node.nsnode) + (adj == ELEM_ADJ_AfterEnd ? 1 : 0));
+        nsIDOMNode_Release(parent);
+        break;
+    case ELEM_ADJ_AfterBegin:
+    case ELEM_ADJ_BeforeEnd:
+        if(is_scopeless_node(elem->node.nsnode))
+            return E_INVALIDARG;
+        hres = mp_set(This, elem->node.nsnode, adj == ELEM_ADJ_AfterBegin ? 0 : get_child_count(elem->node.nsnode));
+        break;
+    default:
+        return E_INVALIDARG;
+    }
+
+    return hres;
 }
 
 static HRESULT WINAPI MarkupPointer2_MoveToPointer(IMarkupPointer2 *iface, IMarkupPointer *pointer)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p)\n", This, pointer);
-    return E_NOTIMPL;
+    MarkupPointer *that;
+    nsIDOMNode *node;
+    LONG off;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p)\n", This, pointer);
+
+    if(!pointer || !(that = unsafe_markup_pointer(pointer)))
+        return E_INVALIDARG;
+    if(that == This)
+        return S_OK;
+    if(!mp_get(that, &node, &off))
+        return E_INVALIDARG;
+
+    hres = mp_set(This, node, off);
+    nsIDOMNode_Release(node);
+    return hres;
 }
 
 static HRESULT WINAPI MarkupPointer2_MoveToContainer(IMarkupPointer2 *iface, IMarkupContainer *container, BOOL at_start)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p %x)\n", This, container, at_start);
-    return E_NOTIMPL;
+    HTMLDocumentNode *doc;
+
+    TRACE("(%p)->(%p %x)\n", This, container, at_start);
+
+    if(!container || !(doc = unsafe_impl_from_IMarkupContainer(container)))
+        return E_INVALIDARG;
+
+    return mp_set(This, doc->node.nsnode, at_start ? 0 : get_child_count(doc->node.nsnode));
+}
+
+/* Moves over one token (or some text) to the right or the left; NoScope elements have neither a
+ * begin nor an end tag. */
+static HRESULT mp_scan(MarkupPointer *This, BOOL right, BOOL move, MARKUP_CONTEXT_TYPE *context,
+        IHTMLElement **element, LONG *len, OLECHAR *text)
+{
+    nsIDOMNode *node, *child, *parent;
+    LONG off, want = len ? *len : 0;
+    MARKUP_CONTEXT_TYPE ctx = CONTEXT_TYPE_None;
+    nsIDOMNode *elem_node = NULL;
+    LONG new_off = 0, copied = 0;
+    nsIDOMNode *new_node = NULL;
+    HRESULT hres = S_OK;
+    BOOL done = FALSE;
+
+    if(element)
+        *element = NULL;
+    if(!mp_get(This, &node, &off))
+        return E_FAIL;
+
+    while(!done) {
+        UINT16 type = get_node_type(node);
+
+        if(type == TEXT_NODE) {
+            LONG tlen = get_text_length(node);
+
+            if(right ? off < tlen : off > 0) {
+                const PRUnichar *str;
+                nsAString nsstr;
+                LONG n = right ? tlen - off : off;
+
+                ctx = CONTEXT_TYPE_Text;
+                if(n > want)
+                    n = want > 0 ? want : 0;
+                if(n && text) {
+                    get_text_node_data(node, &nsstr, &str);
+                    memcpy(text, str + (right ? off : off - n), n * sizeof(WCHAR));
+                    nsAString_Finish(&nsstr);
+                }
+                copied = n;
+                new_node = node;
+                nsIDOMNode_AddRef(new_node);
+                new_off = right ? off + n : off - n;
+                done = TRUE;
+            }else {
+                parent = NULL;
+                nsIDOMNode_GetParentNode(node, &parent);
+                if(!parent)
+                    break;
+                off = get_child_index(parent, node) + (right ? 1 : 0);
+                nsIDOMNode_Release(node);
+                node = parent;
+            }
+            continue;
+        }
+
+        if(right) {
+            if(off < (LONG)get_child_count(node)) {
+                child = get_child_node(node, off);
+                type = get_node_type(child);
+                if(type == TEXT_NODE) {
+                    nsIDOMNode_Release(node);
+                    node = child;
+                    off = 0;
+                }else if(type != ELEMENT_NODE) {
+                    nsIDOMNode_Release(child);
+                    off++;
+                }else if(is_scopeless_node(child)) {
+                    ctx = CONTEXT_TYPE_NoScope;
+                    elem_node = child;
+                    new_node = node;
+                    nsIDOMNode_AddRef(new_node);
+                    new_off = off + 1;
+                    done = TRUE;
+                }else {
+                    ctx = CONTEXT_TYPE_EnterScope;
+                    elem_node = child;
+                    nsIDOMNode_AddRef(child);
+                    new_node = child;
+                    new_off = 0;
+                    done = TRUE;
+                }
+            }else {
+                parent = NULL;
+                if(is_container_node(node))
+                    break;
+                nsIDOMNode_GetParentNode(node, &parent);
+                if(!parent)
+                    break;
+                ctx = CONTEXT_TYPE_ExitScope;
+                elem_node = node;
+                nsIDOMNode_AddRef(elem_node);
+                new_node = parent;
+                new_off = get_child_index(parent, node) + 1;
+                done = TRUE;
+            }
+        }else {
+            if(off > 0) {
+                child = get_child_node(node, off - 1);
+                type = get_node_type(child);
+                if(type == TEXT_NODE) {
+                    nsIDOMNode_Release(node);
+                    node = child;
+                    off = get_text_length(child);
+                }else if(type != ELEMENT_NODE) {
+                    nsIDOMNode_Release(child);
+                    off--;
+                }else if(is_scopeless_node(child)) {
+                    ctx = CONTEXT_TYPE_NoScope;
+                    elem_node = child;
+                    new_node = node;
+                    nsIDOMNode_AddRef(new_node);
+                    new_off = off - 1;
+                    done = TRUE;
+                }else {
+                    ctx = CONTEXT_TYPE_ExitScope;
+                    elem_node = child;
+                    nsIDOMNode_AddRef(child);
+                    new_node = child;
+                    new_off = get_child_count(child);
+                    done = TRUE;
+                }
+            }else {
+                parent = NULL;
+                if(is_container_node(node))
+                    break;
+                nsIDOMNode_GetParentNode(node, &parent);
+                if(!parent)
+                    break;
+                ctx = CONTEXT_TYPE_EnterScope;
+                elem_node = node;
+                nsIDOMNode_AddRef(elem_node);
+                new_node = parent;
+                new_off = get_child_index(parent, node);
+                done = TRUE;
+            }
+        }
+    }
+
+    nsIDOMNode_Release(node);
+
+    if(context)
+        *context = ctx;
+    if(len)
+        *len = copied;
+    if(elem_node) {
+        if(element)
+            hres = mp_node_to_element(elem_node, element);
+        nsIDOMNode_Release(elem_node);
+    }
+
+    if(new_node) {
+        if(SUCCEEDED(hres) && move)
+            hres = mp_set(This, new_node, new_off);
+        nsIDOMNode_Release(new_node);
+    }
+    return hres;
 }
 
 static HRESULT WINAPI MarkupPointer2_Left(IMarkupPointer2 *iface, BOOL move, MARKUP_CONTEXT_TYPE *context,
                                           IHTMLElement **element, LONG *len, OLECHAR *text)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%x %p %p %p %p)\n", This, move, context, element, len, text);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%x %p %p %p %p)\n", This, move, context, element, len, text);
+
+    return mp_scan(This, FALSE, move, context, element, len, text);
 }
 
 static HRESULT WINAPI MarkupPointer2_Right(IMarkupPointer2 *iface, BOOL move, MARKUP_CONTEXT_TYPE *context,
                                            IHTMLElement **element, LONG *len, OLECHAR *text)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%x %p %p %p %p)\n", This, move, context, element, len, text);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%x %p %p %p %p)\n", This, move, context, element, len, text);
+
+    return mp_scan(This, TRUE, move, context, element, len, text);
 }
 
 static HRESULT WINAPI MarkupPointer2_CurrentScope(IMarkupPointer2 *iface, IHTMLElement **p)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    nsIDOMNode *node, *parent;
+    LONG off;
+    HRESULT hres = S_OK;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    *p = NULL;
+    if(!mp_get(This, &node, &off))
+        return E_FAIL;
+
+    /* the element a position is in; a text position is in its parent */
+    if(get_node_type(node) == TEXT_NODE) {
+        parent = NULL;
+        nsIDOMNode_GetParentNode(node, &parent);
+        nsIDOMNode_Release(node);
+        node = parent;
+    }
+    if(node && get_node_type(node) == ELEMENT_NODE)
+        hres = mp_node_to_element(node, p);
+    if(node)
+        nsIDOMNode_Release(node);
+    return hres;
+}
+
+static HRESULT mp_compare_with(MarkupPointer *This, IMarkupPointer *that_iface, int *cmp)
+{
+    MarkupPointer *that;
+
+    if(!that_iface || !(that = unsafe_markup_pointer(that_iface)))
+        return E_INVALIDARG;
+    if(!This->range || !that->range)
+        return E_FAIL;
+    return mp_compare(This, that, cmp);
 }
 
 static HRESULT WINAPI MarkupPointer2_IsLeftOf(IMarkupPointer2 *iface, IMarkupPointer *that_pointer, BOOL *p)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p %p)\n", This, that_pointer, p);
-    return E_NOTIMPL;
+    int cmp;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p %p)\n", This, that_pointer, p);
+
+    hres = mp_compare_with(This, that_pointer, &cmp);
+    if(SUCCEEDED(hres))
+        *p = cmp < 0;
+    return hres;
 }
 
 static HRESULT WINAPI MarkupPointer2_IsLeftOfOrEqualTo(IMarkupPointer2 *iface, IMarkupPointer *that_pointer, BOOL *p)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p %p)\n", This, that_pointer, p);
-    return E_NOTIMPL;
+    int cmp;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p %p)\n", This, that_pointer, p);
+
+    hres = mp_compare_with(This, that_pointer, &cmp);
+    if(SUCCEEDED(hres))
+        *p = cmp <= 0;
+    return hres;
 }
 
 static HRESULT WINAPI MarkupPointer2_IsRightOf(IMarkupPointer2 *iface, IMarkupPointer *that_pointer, BOOL *p)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p %p)\n", This, that_pointer, p);
-    return E_NOTIMPL;
+    int cmp;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p %p)\n", This, that_pointer, p);
+
+    hres = mp_compare_with(This, that_pointer, &cmp);
+    if(SUCCEEDED(hres))
+        *p = cmp > 0;
+    return hres;
 }
 
 static HRESULT WINAPI MarkupPointer2_IsRightOfOrEqualTo(IMarkupPointer2 *iface, IMarkupPointer *that_pointer, BOOL *p)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p %p)\n", This, that_pointer, p);
-    return E_NOTIMPL;
+    int cmp;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p %p)\n", This, that_pointer, p);
+
+    hres = mp_compare_with(This, that_pointer, &cmp);
+    if(SUCCEEDED(hres))
+        *p = cmp >= 0;
+    return hres;
 }
 
 static HRESULT WINAPI MarkupPointer2_IsEqualTo(IMarkupPointer2 *iface, IMarkupPointer *that_pointer, BOOL *p)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p %p)\n", This, that_pointer, p);
-    return E_NOTIMPL;
+    int cmp;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p %p)\n", This, that_pointer, p);
+
+    hres = mp_compare_with(This, that_pointer, &cmp);
+    if(SUCCEEDED(hres))
+        *p = !cmp;
+    return hres;
 }
 
+static HRESULT mp_move_points(MarkupPointer *This, BOOL (*mover)(rangepoint_t*, LONG), LONG cnt)
+{
+    nsIDOMNode *node;
+    rangepoint_t pt;
+    LONG off;
+    HRESULT hres;
+
+    if(!mp_get(This, &node, &off))
+        return E_FAIL;
+    init_rangepoint_no_addref(&pt, node, off);
+    mover(&pt, cnt);
+    hres = mp_set(This, pt.node, pt.off);
+    free_rangepoint(&pt);
+    return hres;
+}
+
+static BOOL mp_mover_chars(rangepoint_t *pt, LONG cnt) { return move_by_chars(pt, cnt) != 0; }
+static BOOL mp_mover_words(rangepoint_t *pt, LONG cnt) { return move_by_words(pt, cnt) != 0; }
+
 static HRESULT WINAPI MarkupPointer2_MoveUnit(IMarkupPointer2 *iface, MOVEUNIT_ACTION action)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%u)\n", This, action);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%u)\n", This, action);
+
+    switch(action) {
+    case MOVEUNIT_NEXTCHAR:
+    case MOVEUNIT_NEXTCLUSTERBEG:
+    case MOVEUNIT_NEXTCLUSTEREND:
+        return mp_move_points(This, mp_mover_chars, 1);
+    case MOVEUNIT_PREVCHAR:
+    case MOVEUNIT_PREVCLUSTERBEG:
+    case MOVEUNIT_PREVCLUSTEREND:
+        return mp_move_points(This, mp_mover_chars, -1);
+    case MOVEUNIT_NEXTWORDBEG:
+    case MOVEUNIT_NEXTWORDEND:
+    case MOVEUNIT_NEXTPROOFWORD:
+        return mp_move_points(This, mp_mover_words, 1);
+    case MOVEUNIT_PREVWORDBEG:
+    case MOVEUNIT_PREVWORDEND:
+    case MOVEUNIT_PREVPROOFWORD:
+        return mp_move_points(This, mp_mover_words, -1);
+    default:
+        FIXME("unsupported action %u\n", action);
+        return E_NOTIMPL;
+    }
 }
 
 static HRESULT WINAPI MarkupPointer2_FindText(IMarkupPointer2 *iface, OLECHAR *text, DWORD flags,
                                               IMarkupPointer *end_match, IMarkupPointer *end_search)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%s %lx %p %p)\n", This, debugstr_w(text), flags, end_match, end_search);
-    return E_NOTIMPL;
+    MarkupPointer *match_end = NULL, *search_end = NULL;
+    BOOL backward = flags & 1, whole = flags & 2, match_case = flags & 4;
+    found_text_t buf = {NULL, 0, 256};
+    rangepoint_t iter, start;
+    nsIDOMNode *node;
+    DWORD len, i, found = ~0u, n;
+    LONG off;
+    WCHAR c;
+    int cmp;
+    HRESULT hres = S_FALSE;
+
+    TRACE("(%p)->(%s %lx %p %p)\n", This, debugstr_w(text), flags, end_match, end_search);
+
+    len = text ? lstrlenW(text) : 0;
+    if(!len)
+        return E_INVALIDARG;
+    if(end_match && !(match_end = unsafe_markup_pointer(end_match)))
+        return E_INVALIDARG;
+    if(end_search && !(search_end = unsafe_markup_pointer(end_search)))
+        return E_INVALIDARG;
+    if(!mp_get(This, &node, &off))
+        return E_FAIL;
+    if(!(buf.chars = malloc(buf.size * sizeof(WCHAR)))) {
+        nsIDOMNode_Release(node);
+        return E_OUTOFMEMORY;
+    }
+
+    /* the characters from the pointer to where the search ends, in the direction of the search */
+    init_rangepoint_no_addref(&iter, node, off);
+    init_rangepoint(&start, node, off);
+    while((c = backward ? move_prev_char(&iter) : move_next_char(&iter))) {
+        if(search_end && search_end->range) {
+            MarkupPointer tmp = *This;
+            tmp.range = NULL;
+            if(SUCCEEDED(mp_set(&tmp, iter.node, iter.off))) {
+                HRESULT h = mp_compare(&tmp, search_end, &cmp);
+                nsIDOMRange_Release(tmp.range);
+                if(SUCCEEDED(h) && (backward ? cmp < 0 : cmp > 0))
+                    break;
+            }
+        }
+        if(!found_text_add(&buf, c)) {
+            hres = E_OUTOFMEMORY;
+            break;
+        }
+    }
+    free_rangepoint(&iter);
+
+    if(hres == S_FALSE && buf.len >= len) {
+        /* in a backward search the characters are collected last first */
+        for(i = 0; i + len <= buf.len; i++) {
+            DWORD k;
+            for(k = 0; k < len; k++) {
+                WCHAR a = buf.chars[i + k], b = text[backward ? len - 1 - k : k];
+                if(match_case ? a != b : towlower(a) != towlower(b))
+                    break;
+            }
+            if(k < len)
+                continue;
+            if(whole) {
+                /* the characters beside the match, in text order */
+                WCHAR left = backward ? (i + len < buf.len ? buf.chars[i + len] : 0) : (i ? buf.chars[i - 1] : 0);
+                WCHAR right = backward ? (i ? buf.chars[i - 1] : 0) : (i + len < buf.len ? buf.chars[i + len] : 0);
+
+                if(is_word_char(left) || is_word_char(right))
+                    continue;
+            }
+            found = i;
+            break;
+        }
+
+        if(found != ~0u) {
+            rangepoint_t a, b;
+
+            /* walk to the match again to get its two ends: a is where the walk of the search
+             * starts to match, b where it ends */
+            init_rangepoint(&a, start.node, start.off);
+            for(n = 0; n < found; n++)
+                backward ? move_prev_char(&a) : move_next_char(&a);
+            init_rangepoint(&b, a.node, a.off);
+            for(n = 0; n < len; n++)
+                backward ? move_prev_char(&b) : move_next_char(&b);
+            /* the pointer goes to the left end of the match, end_match to the right end */
+            hres = mp_set(This, backward ? b.node : a.node, backward ? b.off : a.off);
+            if(SUCCEEDED(hres) && match_end)
+                hres = mp_set(match_end, backward ? a.node : b.node, backward ? a.off : b.off);
+            if(SUCCEEDED(hres))
+                hres = S_OK;
+            free_rangepoint(&a);
+            free_rangepoint(&b);
+        }
+    }
+
+    free_rangepoint(&start);
+    free(buf.chars);
+    return hres;
 }
 
 static HRESULT WINAPI MarkupPointer2_IsAtWordBreak(IMarkupPointer2 *iface, BOOL *p)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    OLECHAR before = 0, after = 0;
+    MARKUP_CONTEXT_TYPE ctx;
+    LONG n;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    if(!This->range)
+        return E_FAIL;
+
+    n = 1;
+    if(FAILED(mp_scan(This, FALSE, FALSE, &ctx, NULL, &n, &before)) || ctx != CONTEXT_TYPE_Text || !n)
+        before = 0;
+    n = 1;
+    if(FAILED(mp_scan(This, TRUE, FALSE, &ctx, NULL, &n, &after)) || ctx != CONTEXT_TYPE_Text || !n)
+        after = 0;
+
+    /* a word starts or ends where a word character meets something else; the ends of the text count */
+    *p = is_word_char(before) != is_word_char(after);
+    return S_OK;
 }
 
 static HRESULT WINAPI MarkupPointer2_GetMarkupPosition(IMarkupPointer2 *iface, LONG *p)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    MarkupPointer *tmp;
+    IMarkupPointer *tmp_iface;
+    MARKUP_CONTEXT_TYPE ctx;
+    nsIDOMNode *node;
+    LONG off, count = 0, n;
+    OLECHAR buf[128];
+    HRESULT hres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    if(!This->range)
+        return E_FAIL;
+    hres = create_markup_pointer(This->doc, &tmp_iface);
+    if(FAILED(hres))
+        return hres;
+    tmp = unsafe_markup_pointer(tmp_iface);
+    mp_get(This, &node, &off);
+    hres = mp_set(tmp, node, off);
+    nsIDOMNode_Release(node);
+
+    /* every token to the left, the characters of text one each */
+    while(SUCCEEDED(hres)) {
+        n = ARRAY_SIZE(buf);
+        hres = mp_scan(tmp, FALSE, TRUE, &ctx, NULL, &n, buf);
+        if(FAILED(hres) || ctx == CONTEXT_TYPE_None)
+            break;
+        count += ctx == CONTEXT_TYPE_Text ? n : 1;
+    }
+    IMarkupPointer_Release(tmp_iface);
+    if(SUCCEEDED(hres))
+        *p = count;
+    return hres;
 }
 
 static HRESULT WINAPI MarkupPointer2_MoveToMarkupPosition(IMarkupPointer2 *iface, IMarkupContainer *container, LONG mp)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p %ld)\n", This, container, mp);
-    return E_NOTIMPL;
+    MARKUP_CONTEXT_TYPE ctx;
+    OLECHAR buf[128];
+    LONG n;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p %ld)\n", This, container, mp);
+
+    if(mp < 0)
+        return E_INVALIDARG;
+    hres = MarkupPointer2_MoveToContainer(iface, container, TRUE);
+    if(FAILED(hres))
+        return hres;
+
+    while(mp > 0) {
+        n = mp < (LONG)ARRAY_SIZE(buf) ? mp : (LONG)ARRAY_SIZE(buf);
+        hres = mp_scan(This, TRUE, TRUE, &ctx, NULL, &n, buf);
+        if(FAILED(hres))
+            return hres;
+        if(ctx == CONTEXT_TYPE_None)
+            return E_INVALIDARG;    /* past the end of the container */
+        mp -= ctx == CONTEXT_TYPE_Text ? n : 1;
+    }
+    return S_OK;
 }
 
 static HRESULT WINAPI MarkupPointer2_MoveUnitBounded(IMarkupPointer2 *iface, MOVEUNIT_ACTION action, IMarkupPointer *boundary)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%u %p)\n", This, action, boundary);
-    return E_NOTIMPL;
+    MarkupPointer *bound;
+    nsIDOMNode *node;
+    LONG off;
+    int before, after;
+    HRESULT hres;
+
+    TRACE("(%p)->(%u %p)\n", This, action, boundary);
+
+    if(!boundary || !(bound = unsafe_markup_pointer(boundary)))
+        return E_INVALIDARG;
+    if(!This->range || !bound->range)
+        return E_FAIL;
+
+    mp_get(This, &node, &off);
+    hres = mp_compare(This, bound, &before);
+    if(SUCCEEDED(hres))
+        hres = MarkupPointer2_MoveUnit(iface, action);
+    if(SUCCEEDED(hres) && SUCCEEDED(mp_compare(This, bound, &after)) && before && after && before != after) {
+        /* went over the boundary: stop on it */
+        nsIDOMNode *bnode;
+        LONG boff;
+
+        mp_get(bound, &bnode, &boff);
+        hres = mp_set(This, bnode, boff);
+        nsIDOMNode_Release(bnode);
+    }
+    nsIDOMNode_Release(node);
+    return hres;
 }
 
 static HRESULT WINAPI MarkupPointer2_IsInsideURL(IMarkupPointer2 *iface, IMarkupPointer *right, BOOL *p)
@@ -2811,8 +3549,10 @@ static HRESULT WINAPI MarkupPointer2_IsInsideURL(IMarkupPointer2 *iface, IMarkup
 static HRESULT WINAPI MarkupPointer2_MoveToContent(IMarkupPointer2 *iface, IHTMLElement *element, BOOL at_start)
 {
     MarkupPointer *This = impl_from_IMarkupPointer2(iface);
-    FIXME("(%p)->(%p %x)\n", This, element, at_start);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p %x)\n", This, element, at_start);
+
+    return MarkupPointer2_MoveAdjacentToElement(iface, element, at_start ? ELEM_ADJ_AfterBegin : ELEM_ADJ_BeforeEnd);
 }
 
 static const IMarkupPointer2Vtbl MarkupPointer2Vtbl = {
@@ -2848,16 +3588,294 @@ static const IMarkupPointer2Vtbl MarkupPointer2Vtbl = {
     MarkupPointer2_MoveToContent
 };
 
-HRESULT create_markup_pointer(IMarkupPointer **ret)
+HRESULT create_markup_pointer(HTMLDocumentNode *doc, IMarkupPointer **ret)
 {
     MarkupPointer *markup_pointer;
 
-    if(!(markup_pointer = malloc(sizeof(*markup_pointer))))
+    if(!(markup_pointer = calloc(1, sizeof(*markup_pointer))))
         return E_OUTOFMEMORY;
 
     markup_pointer->IMarkupPointer2_iface.lpVtbl = &MarkupPointer2Vtbl;
     markup_pointer->ref = 1;
+    markup_pointer->doc = doc;
+    markup_pointer->gravity = POINTER_GRAVITY_Left;
+    IHTMLDOMNode_AddRef(&doc->node.IHTMLDOMNode_iface);
 
     *ret = (IMarkupPointer*)&markup_pointer->IMarkupPointer2_iface;
     return S_OK;
 }
+
+/* The editing of IMarkupServices, between markup pointers. */
+
+/* a range from two positioned pointers; the start has to come first */
+static HRESULT mp_make_range(IMarkupPointer *start_iface, IMarkupPointer *finish_iface, nsIDOMRange **ret)
+{
+    MarkupPointer *start, *finish;
+    nsIDOMNode *node;
+    nsIDOMRange *range;
+    LONG off;
+    int cmp;
+    HRESULT hres;
+
+    if(!(start = unsafe_markup_pointer(start_iface)) || !(finish = unsafe_markup_pointer(finish_iface)))
+        return E_INVALIDARG;
+    if(!start->range || !finish->range)
+        return E_FAIL;
+
+    hres = mp_compare(start, finish, &cmp);
+    if(FAILED(hres))
+        return hres;
+    if(cmp > 0)
+        return E_INVALIDARG;
+
+    if(NS_FAILED(nsIDOMRange_CloneRange(start->range, &range)))
+        return E_FAIL;
+    mp_get(finish, &node, &off);
+    nsIDOMRange_SetEnd(range, node, off);
+    nsIDOMNode_Release(node);
+    *ret = range;
+    return S_OK;
+}
+
+HRESULT markup_remove(IMarkupPointer *start, IMarkupPointer *finish)
+{
+    nsIDOMRange *range;
+    HRESULT hres;
+
+    hres = mp_make_range(start, finish, &range);
+    if(FAILED(hres))
+        return hres;
+    hres = NS_SUCCEEDED(nsIDOMRange_DeleteContents(range)) ? S_OK : E_FAIL;
+    nsIDOMRange_Release(range);
+    return hres;
+}
+
+HRESULT markup_copy(IMarkupPointer *start, IMarkupPointer *finish, IMarkupPointer *target_iface, BOOL move)
+{
+    MarkupPointer *target = unsafe_markup_pointer(target_iface);
+    nsIDOMDocumentFragment *frag = NULL;
+    nsIDOMRange *range, *at;
+    nsIDOMNode *frag_node;
+    int cmp1, cmp2;
+    nsresult nsres;
+    HRESULT hres;
+
+    if(!target)
+        return E_INVALIDARG;
+    hres = mp_make_range(start, finish, &range);
+    if(FAILED(hres))
+        return hres;
+    if(!target->range) {
+        nsIDOMRange_Release(range);
+        return E_FAIL;
+    }
+
+    /* the target cannot be inside of what is copied */
+    if(SUCCEEDED(mp_compare(unsafe_markup_pointer(start), target, &cmp1)) &&
+       SUCCEEDED(mp_compare(target, unsafe_markup_pointer(finish), &cmp2)) && cmp1 < 0 && cmp2 < 0) {
+        nsIDOMRange_Release(range);
+        return E_INVALIDARG;
+    }
+
+    nsres = move ? nsIDOMRange_ExtractContents(range, &frag) : nsIDOMRange_CloneContents(range, &frag);
+    nsIDOMRange_Release(range);
+    if(NS_FAILED(nsres) || !frag)
+        return E_FAIL;
+
+    nsIDOMDocumentFragment_QueryInterface(frag, &IID_nsIDOMNode, (void**)&frag_node);
+    nsIDOMDocumentFragment_Release(frag);
+    hres = NS_SUCCEEDED(nsIDOMRange_CloneRange(target->range, &at)) ? S_OK : E_FAIL;
+    if(SUCCEEDED(hres)) {
+        hres = NS_SUCCEEDED(nsIDOMRange_InsertNode(at, frag_node)) ? S_OK : E_FAIL;
+        nsIDOMRange_Release(at);
+    }
+    nsIDOMNode_Release(frag_node);
+    return hres;
+}
+
+HRESULT markup_insert_text(const WCHAR *text, LONG cch, IMarkupPointer *target_iface)
+{
+    MarkupPointer *target = unsafe_markup_pointer(target_iface);
+    nsIDOMDocument *nsdoc = NULL;
+    nsIDOMText *nstext = NULL;
+    nsIDOMRange *at;
+    nsIDOMNode *node, *text_node;
+    nsAString str;
+    WCHAR *copy;
+    LONG off;
+    nsresult nsres;
+    HRESULT hres = E_FAIL;
+
+    if(!target)
+        return E_INVALIDARG;
+    if(!text || cch < -1)
+        return E_INVALIDARG;
+    if(!mp_get(target, &node, &off))
+        return E_FAIL;
+
+    if(cch < 0)
+        cch = lstrlenW(text);
+    if(!(copy = malloc((cch + 1) * sizeof(WCHAR)))) {
+        nsIDOMNode_Release(node);
+        return E_OUTOFMEMORY;
+    }
+    memcpy(copy, text, cch * sizeof(WCHAR));
+    copy[cch] = 0;
+
+    if(get_node_type(node) == DOCUMENT_NODE)
+        nsIDOMNode_QueryInterface(node, &IID_nsIDOMDocument, (void**)&nsdoc);
+    else
+        nsIDOMNode_GetOwnerDocument(node, &nsdoc);
+    nsIDOMNode_Release(node);
+
+    if(nsdoc) {
+        nsAString_InitDepend(&str, copy);
+        nsres = nsIDOMDocument_CreateTextNode(nsdoc, &str, &nstext);
+        nsAString_Finish(&str);
+        nsIDOMDocument_Release(nsdoc);
+        if(NS_SUCCEEDED(nsres) && nstext) {
+            nsIDOMText_QueryInterface(nstext, &IID_nsIDOMNode, (void**)&text_node);
+            if(NS_SUCCEEDED(nsIDOMRange_CloneRange(target->range, &at))) {
+                hres = NS_SUCCEEDED(nsIDOMRange_InsertNode(at, text_node)) ? S_OK : E_FAIL;
+                nsIDOMRange_Release(at);
+            }
+            nsIDOMNode_Release(text_node);
+            nsIDOMText_Release(nstext);
+        }
+    }
+    free(copy);
+    return hres;
+}
+
+HRESULT markup_insert_element(IHTMLElement *element, IMarkupPointer *start_iface, IMarkupPointer *finish_iface)
+{
+    MarkupPointer *start = unsafe_markup_pointer(start_iface);
+    HTMLElement *elem = element ? unsafe_impl_from_IHTMLElement(element) : NULL;
+    nsIDOMRange *range;
+    HRESULT hres;
+
+    if(!elem || !start)
+        return E_INVALIDARG;
+
+    if(finish_iface) {
+        hres = mp_make_range(start_iface, finish_iface, &range);
+        if(FAILED(hres))
+            return hres;
+        /* the element takes what is between the pointers; that has to be whole elements and text */
+        hres = NS_SUCCEEDED(nsIDOMRange_SurroundContents(range, elem->node.nsnode)) ? S_OK : E_FAIL;
+    }else {
+        if(!start->range)
+            return E_FAIL;
+        if(NS_FAILED(nsIDOMRange_CloneRange(start->range, &range)))
+            return E_FAIL;
+        hres = NS_SUCCEEDED(nsIDOMRange_InsertNode(range, elem->node.nsnode)) ? S_OK : E_FAIL;
+    }
+    nsIDOMRange_Release(range);
+    return hres;
+}
+
+/* the element goes, what it contained stays */
+HRESULT markup_remove_element(IHTMLElement *element)
+{
+    HTMLElement *elem = element ? unsafe_impl_from_IHTMLElement(element) : NULL;
+    nsIDOMNode *parent = NULL, *child, *removed;
+
+    if(!elem)
+        return E_INVALIDARG;
+    nsIDOMNode_GetParentNode(elem->node.nsnode, &parent);
+    if(!parent)
+        return E_INVALIDARG;
+
+    while((child = get_child_node(elem->node.nsnode, 0))) {
+        nsIDOMNode *inserted = NULL;
+        nsIDOMNode_InsertBefore(parent, child, elem->node.nsnode, &inserted);
+        if(inserted)
+            nsIDOMNode_Release(inserted);
+        nsIDOMNode_Release(child);
+    }
+    removed = NULL;
+    nsIDOMNode_RemoveChild(parent, elem->node.nsnode, &removed);
+    if(removed)
+        nsIDOMNode_Release(removed);
+    nsIDOMNode_Release(parent);
+    return S_OK;
+}
+
+static HRESULT txtrange_get_nsrange(IHTMLTxtRange *iface, nsIDOMRange **ret)
+{
+    if(!iface || iface->lpVtbl != &HTMLTxtRangeVtbl)
+        return E_INVALIDARG;
+    *ret = impl_from_IHTMLTxtRange(iface)->nsrange;
+    return S_OK;
+}
+
+HRESULT markup_pointers_to_range(IHTMLTxtRange *range, IMarkupPointer *start_iface, IMarkupPointer *finish_iface)
+{
+    MarkupPointer *start = unsafe_markup_pointer(start_iface), *finish = unsafe_markup_pointer(finish_iface);
+    nsIDOMRange *nsrange;
+    HRESULT hres;
+
+    if(!start || !finish)
+        return E_INVALIDARG;
+    hres = txtrange_get_nsrange(range, &nsrange);
+    if(FAILED(hres))
+        return hres;
+
+    {
+        nsIDOMNode *start_node;
+        LONG start_off;
+
+        nsIDOMRange_GetStartContainer(nsrange, &start_node);
+        nsIDOMRange_GetStartOffset(nsrange, &start_off);
+        hres = mp_set(start, start_node, start_off);
+        nsIDOMNode_Release(start_node);
+        if(SUCCEEDED(hres)) {
+            nsIDOMRange_GetEndContainer(nsrange, &start_node);
+            nsIDOMRange_GetEndOffset(nsrange, &start_off);
+            hres = mp_set(finish, start_node, start_off);
+            nsIDOMNode_Release(start_node);
+        }
+    }
+    return hres;
+}
+
+HRESULT markup_range_to_pointers(IMarkupPointer *start_iface, IMarkupPointer *finish_iface, IHTMLTxtRange *range)
+{
+    MarkupPointer *start = unsafe_markup_pointer(start_iface), *finish = unsafe_markup_pointer(finish_iface);
+    nsIDOMRange *nsrange;
+    nsIDOMNode *node;
+    LONG off;
+    int cmp;
+    HRESULT hres;
+
+    if(!start || !finish)
+        return E_INVALIDARG;
+    hres = txtrange_get_nsrange(range, &nsrange);
+    if(FAILED(hres))
+        return hres;
+    if(!start->range || !finish->range)
+        return E_FAIL;
+    hres = mp_compare(start, finish, &cmp);
+    if(FAILED(hres))
+        return hres;
+    if(cmp > 0)
+        return E_INVALIDARG;
+
+    mp_get(start, &node, &off);
+    nsIDOMRange_SetStart(nsrange, node, off);
+    nsIDOMNode_Release(node);
+    mp_get(finish, &node, &off);
+    nsIDOMRange_SetEnd(nsrange, node, off);
+    nsIDOMNode_Release(node);
+    return S_OK;
+}
+
+HRESULT markup_is_scoped(IHTMLElement *element, BOOL *scoped)
+{
+    HTMLElement *elem = unsafe_impl_from_IHTMLElement(element);
+
+    if(!elem)
+        return E_INVALIDARG;
+    *scoped = !is_scopeless_node(elem->node.nsnode);
+    return S_OK;
+}
